src/lib/gssapi/krb5/set_context_flags.c

```diff
@@ -0,0 +1,41 @@
+/* -*- mode: c; c-basic-offset: 4; indent-tabs-mode: nil -*- */
+
+#include "k5-int.h"
+#include "gssapiP_krb5.h"
+
+#include <stdlib.h>
+
+OM_uint32 KRB5_CALLCONV
+krb5_gss_set_context_flags(OM_uint32 *minor_status, gss_ctx_id_t context,
+                           uint64_t req_flags, uint64_t ret_flags)
+{
+    krb5_gss_ctx_id_t external_context;
+
+    if (context == GSS_C_NO_CONTEXT)
+        return GSS_S_FAILURE | GSS_S_NO_CONTEXT;
+
+    if (minor_status != NULL)
+        *minor_status = 0;
+
+    external_context = (krb5_gss_ctx_id_t)context;
+    if (external_context->magic != KG_CONTEXT)
+        return GSS_S_FAILURE | GSS_S_NO_CONTEXT;
+
+    external_context->gss_flags = req_flags & (GSS_C_CONF_FLAG |
+                                               GSS_C_INTEG_FLAG |
+                                               GSS_C_MUTUAL_FLAG |
+                                               GSS_C_REPLAY_FLAG |
+                                               GSS_C_SEQUENCE_FLAG |
+                                               GSS_C_DELEG_FLAG |
+                                               GSS_C_DCE_STYLE |
+                                               GSS_C_IDENTIFY_FLAG |
+                                               GSS_C_EXTENDED_ERROR_FLAG);
+    external_context->gss_flags |= GSS_C_TRANS_FLAG;
+    if (req_flags & GSS_C_DCE_STYLE)
+        external_context->gss_flags |= GSS_C_MUTUAL_FLAG;
+
+    external_context->req_flags = req_flags;
```

Comment:
Could you specify in the draft what is supposed to happen with req_flags? Currently there is no language specifying what happens in the case of accept_sec_context(). There are technically four sets of flags: set_context_flags's req_flags, set_context_flags's ret_flags_understood, flags from an init_sec_context() sent on the wire, and the nebulous "actual negotiated flags/features used in this connect" (which likely lean towards init's wire flags versus the set_context_flags's flags). In the case of setting the CB flag for psuedo-mech negotiation, that flag's behavior is understood; however, in all other cases, that behavior is unspecified. (e.g., what happens if MUTUAL_FLAG is specified on accept() side, but not on init() side?).